planner/core/rule_join_reorder_greedy.go

```diff
@@ -0,0 +1,203 @@
+// Copyright 2018 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package core
+
+import (
+	"math"
+	"sort"
+
+	"github.com/pingcap/parser/ast"
+	"github.com/pingcap/tidb/expression"
+	"github.com/pingcap/tidb/sessionctx"
+)
+
+func extractInnerJoinGroup(p *LogicalJoin) ([]LogicalPlan, []*expression.ScalarFunction, []expression.Expression) {
+	if p.reordered || p.preferJoinType > uint(0) || p.JoinType != InnerJoin || p.StraightJoin {
+		return nil, nil, nil
+	}
+	lChild := p.children[0]
+	rChild := p.children[1]
+	lJoin, ok := lChild.(*LogicalJoin)
+	if !ok {
+		eqConds := make([]*expression.ScalarFunction, len(p.EqualConditions))
+		otherConds := make([]expression.Expression, len(p.OtherConditions))
+		copy(eqConds, p.EqualConditions)
+		copy(otherConds, p.OtherConditions)
```

Comment:
If `p.EqualCondition` or `p.OtherCondition`'s cap is more than its len, directly return it and then the `append` in line 43 may change the `p.XXXCondition` itself. This copy just want to keep its value not changed, don't copy it is also ok.

For line 114. During `connectedAndNewNode`, we modified the `remainOtherConds`, if we directly use `s.otherConds` it may cause problem because `s.otherConds` should not change in Loop of line 84. But `eqConds` is not changed in `connectedAndNewNode` so we don't need to copy it.
The copy in line 114 can be removed if we check the `OtherConditions` in the following way:
```
the columns in this condition is not all from the left or right, and they can be totally covered by the left+right
```
Im not sure which one is better, but `copy it` needs less code so i choose this one.